test/filters/test_grid_filter.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Point CLoud Library (PCL) - www.pointclouds.org
+ * Copyright (c) 2020-, Open Perception
+ *
+ * All rights reserved
+ */
+
+#include <pcl/filters/experimental/voxel_grid.h>
+#include <pcl/filters/voxel_grid.h>
+#include <pcl/io/pcd_io.h>
+#include <pcl/test/gtest.h>
+#include <pcl/point_types.h>
+
+#include <cmath>
+
+using namespace pcl;
+using namespace pcl::io;
+using namespace Eigen;
+
+PointCloud<PointXYZ>::Ptr cloud(new PointCloud<PointXYZ>);
+PointCloud<PointXYZRGB>::Ptr cloud_organized(new PointCloud<PointXYZRGB>);
+
+const float PRECISION = Eigen::NumTraits<float>::dummy_precision() * 2;
+
+// helper function
+template <typename PointT>
+bool
+isPointApprox(const PointT& p1, const PointT& p2, const float precision)
+{
+  return sqrt(pow(p1.x - p2.x, 2) + pow(p1.y - p2.y, 2) + pow(p1.z - p2.z, 2)) <=
+         precision;
+}
+
+TEST(VoxelGridEquivalentTest, GridFilters)
+{
+  PointCloud<PointXYZ> new_out_cloud, old_out_cloud;
+
+  pcl::experimental::VoxelGrid<PointXYZ> new_grid;
+  new_grid.setLeafSize(0.02f, 0.02f, 0.02f);
+  new_grid.setInputCloud(cloud);
+  new_grid.filter(new_out_cloud);
+
+  pcl::VoxelGrid<PointXYZ> old_grid;
+  old_grid.setLeafSize(0.02f, 0.02f, 0.02f);
+  old_grid.setInputCloud(cloud);
+  old_grid.filter(old_out_cloud);
+
+  EXPECT_EQ(new_out_cloud.size(), old_out_cloud.size());
+  EXPECT_EQ(new_out_cloud.width, old_out_cloud.width);
+  EXPECT_EQ(new_out_cloud.height, old_out_cloud.height);
+
+  for (size_t i = 0; i < new_out_cloud.size(); ++i) {
+    bool same_pt_found = false;
+    for (size_t j = 0; j < old_out_cloud.size(); ++j) {
+      if (isPointApprox(new_out_cloud[i], old_out_cloud[j], PRECISION)) {
+        same_pt_found = true;
+        break;
+      }
+    }
```

Comment:
Can we do better than N^2?

Reply:
Maybe:
1. Sorting based on point coordinate
2. Hash table & rounding off the coordinate to lower precision (b/c numeric error)

Which one do you think is better? Or any better idea? @kunaltyagi 

_Background: the order of the output points is different because the order of iterating grid is different now_

Reply:
Either is fine to be honest.

Reply:
Done by sorting